R/create_integration_matrix_agetime_lag.R

```diff
@@ -0,0 +1,139 @@
+  
+  # Shifting time points by the time caps
+  dat$time1_cap2 <- dat[[time1]] - timecaps[1] + 1
+  dat$time2_cap2 <- dat[[time2]] - timecaps[1] + 1
+  
+  # Number of dimensions in the hazard function
+  if (is.null(Ntime)) Ntime <- max(dat[, "time1_cap", drop = TRUE])
+  if (is.null(Nage)) Nage <- max(dat[age])
+  if (!is.null(strat) & is.null(Nstrat)) Nstrat <- max(dat[strat])
+  # Subsetting data if necessary
+  if (!is.null(subset)) {
+    dat <- subset(dat, eval(parse(text = subset)))
+  }
+  # Number of rows in the resulting matrix
+  nrow <- nrow(dat)
+  # Adding dummy variable for the rows of the matrix
+  dat$row <- seq_len(nrow(dat))
+  
+  # Matrix for 3D hazard function if strat not NULL
+  if (is.null(strat)) {
+    
+    # column entries for integration matrix
+    cols <- apply(dat, 1, function(x) {
+      # If circumcised at birth select relevant entry
+      if (as.numeric(x["time1_cap2"]) == (as.numeric(x["time2_cap2"]))) {
+        test <- min(timecaps[2] - timecaps[1] + 1,
+                    max(1, as.numeric(x["time1_cap2"])))
+      } else {
+        # Else just estimate the
+        test <- cumsum(
+          c(
+            max(1, as.numeric(x["time1_cap2"])),
+            Ntime + (as.numeric(x["time1_cap2"]) :
+                       (as.numeric(x["time2_cap2"]) - 1) > 0 &
+                       as.numeric(x["time1_cap2"]) : 
+                       (as.numeric(x["time2_cap2"]) - 1) <= 
+                       timecaps[2] - timecaps[1])
+          )
+        )
+      }
+      test <- test[-length(test)]
+      return(test)
+    })
+    cols <- unlist(cols)
+    
+    # Row entries for integration matrix
+    rows <- apply(dat, 1, function(x) {
+      rep(as.numeric(x["row"]), as.numeric(x[time2]) - as.numeric(x[time1]))
+    })
+    rows <- unlist(rows)
+    
+    ncol <- Ntime * Nage
+  }
+  # Matrix for 3D hazard function if strat not NULL
+  if (!is.null(strat)) {
+    
+    # column entries for integration matrix
+    cols <- apply(dat, 1, FUN = function(x) {
+      # If circumcised at birth select relevant entry
+      if (as.numeric(x["time1_cap2"]) == (as.numeric(x["time2_cap2"]))) {
+        test <- Ntime * Nage * (as.numeric(x[strat]) - 1) +
+          min(timecaps[2] - timecaps[1] + 1,
+              max(1, as.numeric(x["time1_cap2"])))
+      } else {
+        # Else just estimate the ?
+        test <- cumsum(
+          c(
+            Ntime * Nage * (as.numeric(x[strat]) - 1) + 
+              max(1, as.numeric(x["time1_cap2"])),
+            Ntime + (as.numeric(x["time1_cap2"]) :
+                       (as.numeric(x["time2_cap2"]) - 1) > 0 &
+                       as.numeric(x["time1_cap2"]) : 
+                       (as.numeric(x["time2_cap2"]) - 1) <= 
+                       timecaps[2] - timecaps[1])))
+      }
+      test <- test[-length(test)]
+      return(test)
+    })
+    cols <- unlist(cols)
+    
+    # Row entries for integration matrix
+    rows <- apply(dat, 1, function(x) {
```

Comment:
apply will automatically convert types for you in the output. Might be fine in this case, I can see you have some explicit casts to as.numeric below. But can sometimes be a cause of unexpected bugs. I would usually think about looping with 

```
lapply(nrow(dat), function(i) {
  row <- dat[i, ]
  ## Do stuff with the row
})

Reply:
Went with Jeff's solution